src/lib.rs

```diff
@@ -17,8 +20,95 @@ use tempfile::TempDir;
 use tracing::{debug, info, instrument, warn};
 use url::Url;
 
+use hc_utils::WrappedHeaderHash;
+use holochain::conductor::api::AppResponse;
+use holochain::conductor::api::ZomeCall;
+use holochain_zome_types::{
+    zome::{FunctionName, ZomeName},
+    zome_io::ExternInput,
+    SerializedBytes,
+};
 type HappIds = Vec<String>;
 
+pub async fn activate_holo_hosted_happs(core_happ: Happ) -> Result<()> {
+    let list_of_happs: Vec<WrappedHeaderHash> = get_enabled_hosted_happs(core_happ).await?;
+    info!("got back list_of_happs {:?}", list_of_happs);
+    install_holo_hosted_happs(list_of_happs).await?;
+    Ok(())
+}
+
+pub async fn install_holo_hosted_happs(happs: Vec<WrappedHeaderHash>) -> Result<()> {
+    info!("Starting to install....");
+    // iterate through the vec and
+    // Call http://localhost/hpos-holochain-api/install_hosted_happ
+    // for each WrappedHeaderHash to install the hosted_happ
+    let client = reqwest::Client::new();
+    // Note: Tmp preferences
+    let preferences = Preferences {
+        max_fuel_before_invoice: 1.0,
+        max_time_before_invoice: vec![86400, 0],
+        price_compute: 1.0,
+        price_storage: 1.0,
+        price_bandwidth: 1.0,
+    };
+    for happ_id in happs {
+        info!("Installing happ-id {:?}", happ_id);
+        let body = Body {
+            happ_id: happ_id.0.to_string(),
+            preferences: preferences.clone(),
+        };
+        let response = client
+            .post("http://localhost/hpos-holochain-api/install_hosted_happ")
+            .json(&body)
+            .send()
+            .await?;
+        info!("Installed happ-id {:?}", happ_id);
+        info!("Response {:?}", response);
+    }
+    Ok(())
+}
+
+pub async fn get_enabled_hosted_happs(core_happ: Happ) -> Result<Vec<WrappedHeaderHash>> {
```

Comment:
I have an allergic reaction to allocations when coding in Rust. Here we're returning a `Vec` (requires allocation) by calling `.collect` on an iterator and all we're doing with it at the call site is iterating over it. It would avoid that allocation and therefore be marginally more efficient to return an `impl Iterator` instead

That requires changing three things: The return type of this function to
```suggestion
pub async fn get_enabled_hosted_happs(core_happ: Happ) -> Result<impl Iterator<Item=WrapperHeaderHash>> {
```

The return value of this function
```diff
-                     let happ_bundle_ids: Vec<WrappedHeaderHash> =
-                         .collect();
-                     info!("List of happ_ids {:?}", happ_bundle_ids);
+                    let happ_bundle_ids = happ_bundles.into_iter().map(|happ| happ.happ_id);
```

and the argument type of the function that iterates over it

```diff
- pub async fn install_holo_hosted_happs(happs: Vec<WrappedHeaderHash>) -> Result<()> {
+ pub async fn install_holo_hosted_happs(happs: impl Iterator<Item=WrappedHeaderHash>) -> Result<()> {
```

Reply:
The only reason I think I have preferred using `Vec` is because it is easier to debug. It looks like `Debug` is not implemented for `impl Iterator`

Reply:
Returning `impl Trait` means that you're returning some specific type that implements that trait, and you just don't want to say which type. The actual type here in `std::iter::Map<std::vec::IntoIter<WrapperHeaderHash>, F>` where `F` is the type of the closure `|happ| happ.happ_id`

That type does implement `Debug`, so in a context where you have the full type such as within this function, it's already possible to print out the iterator. In another context where the "type" is `impl Iterator`, such as `install_holo_hosted_happs`, one solution would be to add a `Debug` bound: replace `impl Iterator<Item=WrappedHeaderHash>` with `impl Iterator<Item=WrapperHeaderHash> + Debug`

Note that when you debug one of these iterators it would probably only print out the contents that have not yet been consumed.